Makefile

```diff
@@ -1,11 +1,44 @@
+# The makefile for caffe. Pretty hacky.
+# $(info CXX_OBJS = $(CXX_OBJS))
 PROJECT := caffe
+COMMON_FLAGS += -std=c++11
 
-CONFIG_FILE := Makefile.config
-include $(CONFIG_FILE)
+# uncomment to build with debugging symbols
+DEBUG := 1
 
+# uncomment one
+CPU_ONLY := 1
```

Comment:
The hand-crafted makefiles are confusing. My intention was to have three 
device dependent Makefiles:

Makefile.config.CPU
Makefile.config.CUDA
Makefile.config.OpenCL

and allow variables in Makefile.config to select which file should be 
included:

CPU_ONLY := 1 // build CPU only version
USE_CUDA := 1 // build CUDA version
USE_OPENCL := 1 // build OpenCL version

The OpenCL version still requires the CPU_ONLY flag to be set in order 
to prevent running into CUDA code. The reason is that Caffe branches 
between the CPU code and the CUDA code using the preprocessor:

#ifdef CPU_ONLY
   <CPU code>
#else
   <CUDA code>
#endif

and I haven't fixed that everywhere yet.

In the meantime I add support to build the OpenCL backend using cmake 
which works better and compiles faster.

Robert



On 03/30/2015 09:47 AM, jyegerlehner wrote:
>
> In Makefile 
> <https://github.com/BVLC/caffe/pull/2195#discussion_r27407792>:
>
> >
> > +# uncomment one
> > +CPU_ONLY := 1
>
> In the various Makefile.config.* files, CPU_ONLY := 1 is shown. This 
> is confusing; what is the intended effect if CPU_ONLY is 1 in some and 
> 0 in others? I think there should be a single authoritative source 
> where the decision is made. I'm not sure how all this hangs together, 
> but I'm thinking that since you introduced USE_OPENCL and USE_CUDA, 
> CPU_ONLY should be removed. If neither USE_OPENCL nor USE_CUDA is 
> specified, then what was called CPU_ONLY is what you are left with. 
> I.e. one always gets the CPU implementation. USE_OPENCL and USE_CUDA 
> will determine whether those implementations are available. But maybe 
> you had something else in mind.
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/BVLC/caffe/pull/2195/files#r27407792>.
>